servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/NewRoundRobinLoadBalancer.java

```diff
@@ -246,63 +247,95 @@ public void onSubscribe(final Subscription s) {
 
         @Override
         public void onNext(@Nullable final Collection<? extends ServiceDiscovererEvent<ResolvedAddress>> events) {
-            if (events == null) {
-                LOGGER.debug("{}: unexpectedly received null instead of events.", NewRoundRobinLoadBalancer.this);
+            if (events == null || events.isEmpty()) {
+                LOGGER.debug("{}: unexpectedly received null or empty list instead of events.",
+                        NewRoundRobinLoadBalancer.this);
                 return;
             }
-            for (ServiceDiscovererEvent<ResolvedAddress> event : events) {
-                final ServiceDiscovererEvent.Status eventStatus = event.status();
-                LOGGER.debug("{}: received new ServiceDiscoverer event {}. Inferred status: {}.",
-                        NewRoundRobinLoadBalancer.this, event, eventStatus);
 
+            boolean sendReadyEvent;
+            List<Host<ResolvedAddress, C>> nextHosts;
+            for (;;) {
+                // TODO: we have some weirdness in the event that we fail the CAS namely that we can create a host
+                //  that never gets used but is orphaned. It's fine so long as there is nothing to close but that
+                //  guarantee may not always hold in the future.
```

Comment:
I'm not super familiar with the domain here, but this TODO gives me bad vibes.

Reply:
It's not a new problem from this PR and I wanted to document it because the situaion also gives me bad vibes. 😅 
The only way I can see the remedy it is to either keep a queue of hosts that you made in case they need to be discarded and close them or change the concurrency model. I'm inclined to do the latter since currently the cost of a CAS fail is quite high.